core/src/bytes/zarith/zarith.h

```diff
@@ -0,0 +1,160 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+
+#ifndef LEDGER_CORE_ZARITH_H
+#define LEDGER_CORE_ZARITH_H
+
+#include <algorithm>
+#include <vector>
+#include <bytes/BytesReader.h>
+#include <utils/hex.h>
+
+namespace ledger {
+    namespace core {
+        class zarith {
+        public:
+            /*
+             * Most significant bit of each byte is replaced with 1 to let
+             * us know there is still something to read, else 0
+             * To sum up, if the resulting bytes are "chained" together (little endian) and we drop
+             * the most significant bit of each byte, we get the input ("chained") bytes
+             * e.g.
+             * Input:           ad d9                  10101101 11011001  (I)
+             * Output:          d9 db 02      11011001 11011011 00000010
+             * LE:              02 db d9      00000010 11011011 11011001
+             * Drop 1st bit:            (0)0000010 (1)1011011 (1)1011001  == (I)
+            */
+            static std::vector<uint8_t> zSerializeNumber(const std::vector<uint8_t> &inputData) {
+                std::vector<uint8_t> result;
+                size_t id = inputData.size() - 1, delay = 0; // delay is used when offset > 7
+                uint8_t offset = 0;
+                while (id + delay >= 0 && id + delay < inputData.size()) {
+                    auto byte = inputData[id + delay];
+
+                    // We shift current byte
+                    byte <<= offset;
+
+                    // Take shifted bits from previous byte
+                    byte |= id + delay < inputData.size() - 1 ? inputData[id + delay + 1] >> (7 - (offset - 1)) : 0x00;
+
+                    // This boolean will let us know if shifting last byte of inputData
+                    // will generate a new byte (with remaining shifted bytes)
+                    bool isFirst = id + delay == 0;
+                    bool needAdditionalByte = isFirst && inputData[id + delay] >> (7 - offset);
+
+                    // We set first bit to 1 to know that there is a coming byte
+                    if (!isFirst || needAdditionalByte) {
+                        byte |= 0x80;
+                    }
+
+                    // Push to result modified byte
+                    result.push_back(byte);
+                    if (needAdditionalByte) {
+                        result.push_back(inputData[id + delay] >> (7 - offset));
+                    }
+
+                    // Update offset: if offset % 7 == 0 then following byte
+                    // is not shifted i.e. offset = 0
+                    offset = offset > 0 && !(offset % 7) ? 0 : offset + 1;
+
+                    // If we shifted with [id/7] > 7 then we have a "delay"
+                    // meaning that we should read values from "delayed" byte
+                    delay = ((inputData.size() - 1) - id) / 7;
```

Comment:
1. Looks like there is a bug here:
id = 0;
delay  = 0;
inputData.size() = 8;
we just processed inputData[id + delay] -> inputData[0],
now delay = ((8 - 1) - 0) / 7 -> delay became 1, in the next line id became -1, so next iteration we will still proceed inputaData[0] so we processed this byte twice.
2. I think you don't need `delay` at all.

Reply:
1. Actually here the issue is that we should `break` after processing `id + delay == 0`, but regarding your remark, in case `inputData .size() > 8` e.g. `inputData .size() = 9`, for `(id , delay, offset) = (1, 0, 7)`, in the first iteration, `inputData[1]` will be shifted by 7, so basically in `result` you push a byte `1<7 most significant bits from inputData[2]>`, the second iteration, with `(id , delay, offset) = (0, 1, 0)`  will push `x<7 less significant bits from inputData[1]>` (`x` depends if most significant bit of `inputData[1]` is a 0 or 1 that's when we add an additional byte)
2. I think we need `delay` for cases when `inputData.size() > 8` 
We can discuss that on Friday or Monday it will be easier! 😉

Reply:
1. Lets look at this example { 0xFF, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01 };
the code will get output 81 80 80 80 80 80 80 80 **FF FF** 01,  that can't be right - too many new ones.
2. You only use  `inputData[id + delay]` and `inputData[id + delay + 1]`, that can be replaced by `inputData[idx]` and `inputData[idx + 1]` and a new way of changing `idx`.

Reply:
So yeah this is an issue which is explained but my comment above, "Actually here the issue is that we should `break` after processing `id + delay == 0`" 
Btw, I added a test with test vector above ! 👍

Reply:
Thanks for adding the test.
But the bug remains there, and it is not related to the last byte, the problem is that we proceed with  the same bytes twice when we increment `delay` and decrement `id`
Try this vector `{ 0x01, 0x00, 0xFF, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01 }` it will again give the answer with two many "ones".
BTW you use `size_t` for `id`, `id` can became negative, so it start to overflow, there were no errors because when we add `delay` to overflowed `id` it overflows again and became normal number, but I found this dangerous anyway.

Reply:
Oh yeah nice catch for the example ! 👍 
For the overflow thing, if you take into account the `break` condition that I added above and test on `id + delay` we have, it's not dangerous, we care about `id + delay` reason why, I agree, we should replace it by only `id` which is also simplifying the condition to increment `id` 
I think now it should be fine, WDYT ? 🤔 
Thanks for the review BTW ! 🙏

Reply:
For me it looks good now, just few comments that should speed up a little bit and increase readability.